design-history/exploration-reports/2019.06-unixfsv2-spike-01.md

```diff
@@ -0,0 +1,296 @@
+A unixfsv2 spike, using schemas
+===============================
+
+This document is an exploration of what a hypothetical "unixfs v2" might look like,
+using schemas as a design language for describing the data structures.
+
+This is not a final proposal, but rather a lens to explore both unixfsv2
+concepts and schemas themselves at the same time, and how we expect them
+to contribute to each other.  Several concepts in schemas, advanced layouts,
+their interactions, and their syntaxes will be trialled here; they are not final.
+
+Without further ado...
+
+Schema
+------
+
+```
+type AnyFile union {
+	| File "f"
+	| Dir "d"
+	| Symlink "s"
+} representation keyed
+
+type Filename string
+
+type Dir struct {
+	members {Filename:DirEnt}<HAMT>
+}
+
+type DirEnt struct {
+	content &AnyFile
+	attribs Attribs
+}
+
+type File struct {
+	content Bytes<Rabin>
+}
+
+type Symlink struct {
+	target String
+}
+
+advanced Rabin {
+	kind bytes
+	implementation "IPLD/experimental/rabin/v1"
+}
+
+advanced HAMT {
+	kind map
+	implementation "IPLD/experimental/HAMT/v1"
+}
+
+type Attribs struct {
+	mtime Int # Danger danger will robinson; int sizes will rear their ugly head.
+	posix Int # Can we just... do the standard 0777 mask packing here?  I don't think anything else is *less* work.
```

Comment:
A note on this particular point. Thre **is** a way to go with less work, at least on the golang side.
This is how I re-package all of `filemode` into a little endian 32 bit valie on my side for UFS1.5:
```
      # Mode bit-positions are protocol-safe:
      # https://golang.org/pkg/os/#FileMode
      # https://github.com/golang/go/issues/25422#issuecomment-389497051
      ( $_->{mode} & 0777 )
        |
      ( ( S_IFDIR == $_->{ifmt} ) && 1<<31 )
        |
      ( ( S_IFLNK == $_->{ifmt} ) && 1<<27 )
        |
      ( ( $_->{mode} & S_ISUID ) && 1<<23 )
        |
      ( ( $_->{mode} & S_ISGID ) && 1<<22 )
        |
      ( ( $_->{mode} & S_ISVTX ) && 1<<20 )
```

Reply:
Typical Unix systems do with a 16-bit mode value. The top four bits (S_IFMT) are used for seven types of files, the middle three for the three special bits, and the last 9 for the traditional permissions. The only good thing about this stuff (save for the sizes) is that it matches the "100644" notation found in git, but there ain't much else.

Either way, the three bool fields definitely should be stored with the mode.